### PR TITLE
compiling and running h2o on alpine linux

### PR DESCRIPTION
**Pull request by felixvonberlin**
This PR adapts some changes to h2o for compiling h2o as library on alpine linux.

##### `CMakeLists.txt`:
- added automated detection of alpine linux (and therefore the MUSL lib)
- removed _GNU_SOURCE when compiling hiredis (issue #3394)
##### `include/h2o/memory.h`:
- updated macro branch for hiding some GNU-specific (?) code on alpine linux 
##### `include/h2o/memory.c`:
- updated macro branch for hiding some GNU-specific (?) code on alpine linux 

Regards,
Felix

### CODE REVIEW

**Comment by felixvonberlin** [comment]
fyi: I'll investigate the reason for the failed pipeline in two weeks.

**Comment by felixvonberlin** [comment]
One issue in the workflow was that asking CMake for the distribution is only possible since version 3.22. The workflow seems to use ubuntu16.04 with uses an older version.

Another issue is completing `t/40http3-forward.t`:
```
# t/40http3-forward.t --------------------------------
  0.092670 spawning /home/ci/h2o... [/tmp/MaRk_fKqJ1:11] in command quic-nodes, failed to resolve address: Address family for hostname not supported
  0.188741 server failed to start (got 19968)
  0.193344 Command exited with non-zero status 111
  0.193365 0.06user 0.03system 0:00.19elapsed 50%CPU (0avgtext+0avgdata 31104maxresident)k
  0.193415 0inputs+0outputs (0major+9987minor)pagefaults 0swaps
# t/40http3-forward.t failed

```
Until now, I'm quite unsure why that step fails.

**Comment by felixvonberlin** [comment]
I checked again, the test `t/40http3/forward.t` fails for me also on the master branch.
So it's maybe about my machine although this is a docker based test?

**Comment by felixvonberlin** [comment]
@kazuho Do I need to do something on my side?

File: CMakeLists.txt [hunk 1]
```diff
@@ -278,6 +285,12 @@ IF (NOT ARCH_SUPPORTS_64BIT_ATOMICS)
 ENDIF (NOT ARCH_SUPPORTS_64BIT_ATOMICS)
 IF (CMAKE_SYSTEM_NAME STREQUAL "Linux")
     SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -pthread -D_GNU_SOURCE")
+    IF ("${${DISTRO}}" STREQUAL "alpine")
+	file(GLOB SPECIAL_SRC_FILES
+        	"deps/hiredis/*")
+	set_source_files_properties(${SPECIAL_SRC_FILES} PROPERTIES COMPILE_FLAGS -U_GNU_SOURCE)
+	SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -D_MUSL_LIB")
+    ENDIF ()
```
Comment (kazuho) [inline CMakeLists.txt:293]: I think we no longer need this now that #3406 is merged.

File: include/h2o/memory.h [hunk 2]
```diff
@@ -54,7 +54,7 @@ extern "C" {
 #define H2O_GNUC_VERSION 0
 #endif
 
-#if __STDC_VERSION__ >= 201112L
+#if __STDC_VERSION__ >= 201112L && !defined(_MUSL_LIB)
```
Comment (kazuho) [inline include/h2o/memory.h:57]: This is a C compiler issue rather than a libc issue. It has been resolved in #3427.

File: lib/common/memory.c [hunk 3]
```diff
@@ -604,7 +604,7 @@ void h2o_append_to_null_terminated_list(void ***list, void *element)
 
 char *h2o_strerror_r(int err, char *buf, size_t len)
 {
-#if !(defined(_GNU_SOURCE) && defined(__gnu_linux__))
+#if (!(defined(_GNU_SOURCE) && defined(__gnu_linux__))) || defined(_MUSL_LIB)
```
Comment (kazuho) [inline lib/common/memory.c:607]: I do not think this is necessary as IIUC `defined(__gnu_linux__)` evaluates true only if glibc is in use.

**Review by kazuho — COMMENTED** [review]
Thank you for the PR, sorry for the delay.

To me, with all the changes applied since this PR has been opened, the only change necessary is #3450.

**Pull request closed without merging**